compiler/AST/AggregateType.cpp

```diff
@@ -2224,9 +2226,124 @@ void AggregateType::buildDefaultInitializer() {
   }
 }
 
+static bool hasFullyGenericField(AggregateType* at) {
+  for_fields(field, at) {
+    DefExpr* defExpr = field->defPoint;
+    if (!field->hasFlag(FLAG_TYPE_VARIABLE) && !field->hasFlag(FLAG_PARAM) &&
+        !field->hasFlag(FLAG_SUPER_CLASS)) {
+      if (defExpr->exprType == nullptr && defExpr->init == nullptr) {
+        return true;
+      }
+    }
+  }
+
+  if (at->isClass()                    ==  true &&
+      at->symbol->hasFlag(FLAG_REF)    == false &&
+      at->dispatchParents.n            >      0 &&
+      at->symbol->hasFlag(FLAG_EXTERN) == false) {
+    if (AggregateType* parent = at->dispatchParents.v[0]) {
+      return hasFullyGenericField(parent);
+    }
+  }
+
+  return false;
+}
+
+void AggregateType::buildReaderInitializer() {
+  if (!fUseIOFormatters) return;
+  if (this->getModule()->modTag == MOD_INTERNAL) return;
+
+  if (hasFullyGenericField(this)) return;
+
+  if (builtReaderInit == false &&
+      symbol->hasFlag(FLAG_REF) == false) {
+    SET_LINENO(this);
+    FnSymbol*  fn    = new FnSymbol("init");
+    ArgSymbol* _mt   = new ArgSymbol(INTENT_BLANK, "_mt",  dtMethodToken);
+    ArgSymbol* _this = new ArgSymbol(INTENT_BLANK, "this", this);
+
+    ArgSymbol* reader = new ArgSymbol(INTENT_BLANK, "chpl__reader", dtAny);
+    fn->where = new BlockStmt(new CallExpr("chpl__isFileReader",
+                              new CallExpr(PRIM_TYPEOF, reader)));
+
+    fn->cname = fn->name;
+    fn->_this = _this;
+
+    // Lydia NOTE 06/16/17: I don't think I want to add the
+    //  DEFAULT_CONSTRUCTOR flag to this function, but if I do,
+    // then I will need to do something different in wrappers.cpp.
+    fn->addFlag(FLAG_COMPILER_GENERATED);
+    fn->addFlag(FLAG_LAST_RESORT);
+    fn->addFlag(FLAG_SUPPRESS_LVALUE_ERRORS);
+
+    _this->addFlag(FLAG_ARG_THIS);
+
+    fn->insertFormalAtTail(_mt);
+    fn->insertFormalAtTail(_this);
+    fn->insertFormalAtTail(reader);
+
+
+    if (this->isUnion() == false) {
+      std::set<const char*> names;
+      SymbolMap fieldArgMap;
+
+      if (addSuperArgs(fn, names, fieldArgMap, reader) == true) {
+        // Parent fields before child fields
+        fieldToArg(fn, names, fieldArgMap, reader);
+
+        {
+          CallExpr* getFormatter = new CallExpr(".", reader, new_CStringSymbol("formatter"));
+          CallExpr* readField = new CallExpr("readTypeStart", gMethodToken, getFormatter,
+                                             reader, new SymExpr(this->symbol));
+          fn->insertAtHead(readField);
+        }
+        {
+          CallExpr* getFormatter = new CallExpr(".", reader, new_CStringSymbol("formatter"));
```

Comment:
I would imagine that you should store the formatter gotten in a temporary variable rather than computing it multiple times here. You'd probably pass it as an additional argument to `fieldToArg` as well.